File: packages/earthsci_toolkit/src/earthsci_toolkit/canonicalize.py
"""Canonical AST form per discretization RFC §5.4.

Implements ``canonicalize(expr)`` and ``canonical_json(expr)`` such that two
ASTs are canonically equal iff their ``canonical_json`` outputs are
byte-identical.

See ``docs/rfcs/discretization.md`` §5.4.1–§5.4.7 for the normative rules.
"""

from __future__ import annotations

import math
from dataclasses import replace
from typing import Any, List

from .esm_types import Expr, ExprNode


class CanonicalizeError(Exception):
    """Base class for canonicalization errors (RFC §5.4.6 / §5.4.7)."""

    code: str = ""


class NonFiniteError(CanonicalizeError):
    """``E_CANONICAL_NONFINITE`` — NaN or ±Inf encountered (§5.4.6)."""

    code = "E_CANONICAL_NONFINITE"

    def __init__(self, message: str = "E_CANONICAL_NONFINITE"):
        super().__init__(message)


class DivByZeroError(CanonicalizeError):
    """``E_CANONICAL_DIVBY_ZERO`` — ``/(0, 0)`` encountered (§5.4.7)."""

    code = "E_CANONICAL_DIVBY_ZERO"

    def __init__(self, message: str = "E_CANONICAL_DIVBY_ZERO"):
        super().__init__(message)


def canonicalize(expr: Expr) -> Expr:
    """Canonicalize an expression tree per RFC §5.4. Input is not mutated."""
    if isinstance(expr, bool):
        # Treat booleans as integers in JSON terms — but ESM expressions
        # don't legitimately carry booleans; surface as int for safety.
        return int(expr)
    if isinstance(expr, int):
        return expr
    if isinstance(expr, float):
        if not math.isfinite(expr):
            raise NonFiniteError()
        return expr
    if isinstance(expr, str):
        return expr
    if isinstance(expr, ExprNode):
        return _canon_op(expr)
    raise TypeError(f"unknown expression type: {type(expr).__name__}")


def canonical_json(expr: Expr) -> str:
    """Emit canonical on-wire JSON (sorted keys, no whitespace, §5.4.6 numbers)."""
    c = canonicalize(expr)
    return _emit_json(c)


def _canon_op(node: ExprNode) -> Expr:
    new_args: List[Expr] = [canonicalize(a) for a in node.args]
    work = replace(node, args=new_args)
    if work.op == "+":
        return _canon_add(work)
    if work.op == "*":
        return _canon_mul(work)
    if work.op == "-":
        return _canon_sub(work)
    if work.op == "/":
        return _canon_div(work)
    if work.op == "neg":
        return _canon_neg(work)
    return work


def _canon_add(node: ExprNode) -> Expr:
    flat = _flatten_same_op(node.args, "+")
    others, _had_int_zero, had_float_zero = _partition_identity(flat, 0)
    if had_float_zero and not _all_float_literals(others):
        others.append(0.0)
    if not others:
        return 0.0 if had_float_zero else 0
    if len(others) == 1:
        return others[0]
    _sort_args(others)
    return replace(node, op="+", args=others)


def _canon_mul(node: ExprNode) -> Expr:
    flat = _flatten_same_op(node.args, "*")
    for a in flat:
        if isinstance(a, int) and not isinstance(a, bool) and a == 0:
            return 0
        if isinstance(a, float) and a == 0.0:
            # Preserve signbit (multiplying by 0.0 does not flip sign).
            return a * 0.0
    others, _had_int_one, had_float_one = _partition_identity(flat, 1)
    if had_float_one and not _all_float_literals(others):
        others.append(1.0)
    if not others:
        return 1.0 if had_float_one else 1
    if len(others) == 1:
        return others[0]
    _sort_args(others)
    return replace(node, op="*", args=others)


def _canon_sub(node: ExprNode) -> Expr:
    if len(node.args) == 1:
        return _canon_neg_value(node.args[0])
    if len(node.args) == 2:
        a, b = node.args[0], node.args[1]
        if _is_zero_any(a):
            return _canon_neg_value(b)
        if _is_zero_any(b):
            if isinstance(b, float) and isinstance(a, int) and not isinstance(a, bool):
                return float(a)
            return a
    return node


def _canon_div(node: ExprNode) -> Expr:
    if len(node.args) != 2:
        return node
    a, b = node.args[0], node.args[1]
    if _is_zero_any(a) and _is_zero_any(b):
        raise DivByZeroError()
    if _is_one_any(b):
        if isinstance(b, float) and isinstance(a, int) and not isinstance(a, bool):
            return float(a)
        return a
    if _is_zero_any(a):
        return 0.0 if isinstance(a, float) else 0
    return node


def _canon_neg(node: ExprNode) -> Expr:
    if len(node.args) != 1:
        return node
    return _canon_neg_value(node.args[0])


def _canon_neg_value(arg: Expr) -> Expr:
    if isinstance(arg, bool):
        arg = int(arg)
    if isinstance(arg, int):
        return -arg
    if isinstance(arg, float):
        return -arg
    if isinstance(arg, ExprNode) and arg.op == "neg" and len(arg.args) == 1:
        return arg.args[0]
    return ExprNode(op="neg", args=[arg])


def _flatten_same_op(args: List[Expr], op: str) -> List[Expr]:
    out: List[Expr] = []
    for a in args:
        if isinstance(a, ExprNode) and a.op == op:
            out.extend(a.args)
        else:
            out.append(a)
    return out


def _partition_identity(args: List[Expr], identity: int):
    others: List[Expr] = []
    had_int = False
    had_float = False
    for a in args:
        if isinstance(a, bool):
            others.append(int(a))
            continue
        if isinstance(a, int) and a == identity:
            had_int = True
            continue
        if isinstance(a, float) and a == float(identity):
            had_float = True
            continue
        others.append(a)
    return others, had_int, had_float


def _all_float_literals(args: List[Expr]) -> bool:
    return bool(args) and all(isinstance(a, float) for a in args)


def _is_zero_any(e: Expr) -> bool:
    if isinstance(e, bool):
        return False
    return (isinstance(e, int) and e == 0) or (isinstance(e, float) and e == 0.0)


def _is_one_any(e: Expr) -> bool:
    if isinstance(e, bool):
        return False
    return (isinstance(e, int) and e == 1) or (isinstance(e, float) and e == 1.0)


def _arg_tier(e: Expr) -> int:
    if isinstance(e, (int, float)) and not isinstance(e, bool):
        return 0
    if isinstance(e, str):
        return 1
    if isinstance(e, ExprNode):
        return 2
    return 3


def _numeric_key(e: Expr) -> float:
    if isinstance(e, bool):
        return 0.0
    if isinstance(e, int):
        return float(e)
    if isinstance(e, float):
        return e
    return 0.0


def _sort_args(args: List[Expr]) -> None:
    # Memoize canonical JSON for non-leaf nodes (§5.4.9).
    cache: dict[int, str] = {}

    def get_json(idx: int, e: Expr) -> str:
        if idx not in cache:
            cache[idx] = _emit_json(e)
        return cache[idx]

    indices = list(range(len(args)))

    def sort_key(idx: int):
        e = args[idx]
        tier = _arg_tier(e)
        if tier == 0:
            return (tier, _numeric_key(e), isinstance(e, float))
        if tier == 1:
            return (tier, e)
        return (tier, get_json(idx, e))

    indices.sort(key=sort_key)
    snap = [args[i] for i in indices]
    for i, v in enumerate(snap):
        args[i] = v


def _emit_json(e: Expr) -> str:
    if isinstance(e, bool):
        return "true" if e else "false"
    if isinstance(e, int):
        return str(e)
    if isinstance(e, float):
        if not math.isfinite(e):
            raise NonFiniteError()
        return format_canonical_float(e)
    if isinstance(e, str):
        return _json_string(e)
    if isinstance(e, ExprNode):
        return _emit_node_json(e)
    if e is None:
        return "null"
    raise TypeError(f"cannot canonicalize value of type {type(e).__name__}")


def _emit_node_json(n: ExprNode) -> str:
    entries: list[tuple[str, str]] = []
    entries.append(("op", _json_string(n.op)))
    args_str = "[" + ",".join(_emit_json(a) for a in n.args) + "]"
    entries.append(("args", args_str))
    if n.wrt is not None:
        entries.append(("wrt", _json_string(n.wrt)))
    if n.dim is not None:
        entries.append(("dim", _json_string(n.dim)))
    if getattr(n, "handler_id", None) is not None:
        entries.append(("handler_id", _json_string(n.handler_id)))
    entries.sort(key=lambda kv: kv[0])
    body = ",".join(f"{_json_string(k)}:{v}" for k, v in entries)
    return "{" + body + "}"


def _json_string(s: str) -> str:
    # Use json.dumps for proper string escaping (ensures \uXXXX where needed).
    import json

    return json.dumps(s, ensure_ascii=False)


def format_canonical_float(f: float) -> str:
    """Format a finite ``float`` per RFC §5.4.6."""
    if not math.isfinite(f):
        raise NonFiniteError()
    if f == 0.0:
        return "-0.0" if math.copysign(1.0, f) < 0 else "0.0"
    abs_f = abs(f)
    use_exp = abs_f < 1e-6 or abs_f >= 1e21
    if use_exp:
        # Python's repr() picks shortest round-trip and may use exponent form
        # already; format with %e and re-normalize.
        # Get shortest round-trip mantissa and exponent.
        s = f"{f!r}"
        if "e" not in s and "E" not in s:
            # Force exponent form.
            s = f"{f:.17e}"
            s = _trim_mantissa(s)
        else:
            s = s.lower().replace("e+", "e")
        return _normalize_exponent(s)
    # Plain decimal range.
    s = f"{f!r}"
    if "e" in s or "E" in s:
        # Convert exponent form to plain decimal.
        s = _expand_to_plain(f)
    if "." not in s:
        s += ".0"
    return s


def _trim_mantissa(s: str) -> str:
    """Trim trailing zeros in the mantissa of an exponent-form string."""
    if "e" not in s and "E" not in s:
        return s
    s = s.lower()
    mant, exp = s.split("e", 1)
    if "." in mant:
        mant = mant.rstrip("0").rstrip(".")
    return f"{mant}e{exp}"


def _expand_to_plain(f: float) -> str:
    # Use Python's ``f`` format and rstrip useless zeros while preserving precision.
    # Determine number of decimals needed via repr.
    s = f"{f!r}"
    if "e" not in s.lower():
        return s
    # Use general format with high precision and strip.
    # Prefer Decimal for clean conversion.
    from decimal import Decimal

    d = Decimal(repr(f))
    txt = format(d, "f")
    # Trim trailing zeros after decimal but keep at least "X.0".
    if "." in txt:
        txt = txt.rstrip("0").rstrip(".")
    return txt


def _normalize_exponent(s: str) -> str:
    if "e" not in s:
        return s
    mant, exp = s.split("e", 1)
    if exp.startswith("+"):
        exp = exp[1:]
    sign = ""
    if exp.startswith("-"):
        sign = "-"
        exp = exp[1:]
    exp = exp.lstrip("0") or "0"
    return f"{mant}e{sign}{exp}"

File: packages/EarthSciSerialization.jl/src/canonicalize.jl
# Canonical AST form per discretization RFC §5.4.
#
# Implements `canonicalize(expr)` and `canonical_json(expr)` such that two
# ASTs are canonically equal iff their `canonical_json` outputs are
# byte-identical.
#
# See `docs/rfcs/discretization.md` §5.4.1–§5.4.7 for the normative rules.

"""
    CanonicalizeError(code::String, message::String)

Error raised by [`canonicalize`](@ref). The `code` field carries one of the
RFC §5.4.6 / §5.4.7 stable error codes (`E_CANONICAL_NONFINITE`,
`E_CANONICAL_DIVBY_ZERO`).
"""
struct CanonicalizeError <: Exception
    code::String
    message::String
end

Base.showerror(io::IO, e::CanonicalizeError) =
    print(io, "CanonicalizeError(", e.code, "): ", e.message)

"""
    canonicalize(expr::Expr) -> Expr

Canonicalize an expression tree per discretization RFC §5.4. Returns a new
tree; the input is not mutated. Throws [`CanonicalizeError`](@ref) for
NaN/Inf or `0/0`.
"""
function canonicalize end

canonicalize(e::IntExpr) = e

function canonicalize(e::NumExpr)
    isfinite(e.value) || throw(CanonicalizeError("E_CANONICAL_NONFINITE",
        "non-finite float in canonical form"))
    return e
end

canonicalize(e::VarExpr) = e

function canonicalize(e::OpExpr)
    new_args = Vector{Expr}(undef, length(e.args))
    for (i, a) in enumerate(e.args)
        new_args[i] = canonicalize(a)
    end
    work = OpExpr(e.op, new_args;
                  wrt=e.wrt, dim=e.dim, output_idx=e.output_idx,
                  expr_body=e.expr_body, reduce=e.reduce, ranges=e.ranges,
                  regions=e.regions, values=e.values, shape=e.shape,
                  perm=e.perm, axis=e.axis, fn=e.fn,
                  name=e.name, value=e.value)
    if work.op == "+"
        return _canon_add(work)
    elseif work.op == "*"
        return _canon_mul(work)
    elseif work.op == "-"
        return _canon_sub(work)
    elseif work.op == "/"
        return _canon_div(work)
    elseif work.op == "neg"
        return _canon_neg(work)
    end
    return work
end

function _canon_add(node::OpExpr)
    flat = _flatten_same_op(node.args, "+")
    others, _had_int_zero, had_float_zero = _partition_identity(flat, 0)
    if had_float_zero && !_all_float_literals(others)
        push!(others, NumExpr(0.0))
    end
    if isempty(others)
        return had_float_zero ? NumExpr(0.0) : IntExpr(0)
    end
    if length(others) == 1
        return others[1]
    end
    _sort_args!(others)
    return OpExpr("+", others)
end

function _canon_mul(node::OpExpr)
    flat = _flatten_same_op(node.args, "*")
    for a in flat
        if a isa IntExpr && a.value == 0
            return IntExpr(0)
        end
        if a isa NumExpr && a.value == 0.0
            # Preserve signbit: -0.0 stays as -0.0 via copysign.
            return NumExpr(copysign(0.0, a.value))
        end
    end
    others, _had_int_one, had_float_one = _partition_identity(flat, 1)
    if had_float_one && !_all_float_literals(others)
        push!(others, NumExpr(1.0))
    end
    if isempty(others)
        return had_float_one ? NumExpr(1.0) : IntExpr(1)
    end
    if length(others) == 1
        return others[1]
    end
    _sort_args!(others)
    return OpExpr("*", others)
end

function _canon_sub(node::OpExpr)
    if length(node.args) == 1
        return _canon_neg_value(node.args[1])
    end
    if length(node.args) == 2
        a, b = node.args[1], node.args[2]
        if _is_zero_any(a)
            return _canon_neg_value(b)
        end
        if _is_zero_any(b)
            if b isa NumExpr && a isa IntExpr
                return NumExpr(Float64(a.value))
            end
            return a
        end
    end
    return node
end

function _canon_div(node::OpExpr)
    length(node.args) == 2 || return node
    a, b = node.args[1], node.args[2]
    if _is_zero_any(a) && _is_zero_any(b)
        throw(CanonicalizeError("E_CANONICAL_DIVBY_ZERO", "0/0 in canonical form"))
    end
    if _is_one_any(b)
        if b isa NumExpr && a isa IntExpr
            return NumExpr(Float64(a.value))
        end
        return a
    end
    if _is_zero_any(a)
        return a isa NumExpr ? NumExpr(0.0) : IntExpr(0)
    end
    return node
end

function _canon_neg(node::OpExpr)
    length(node.args) == 1 || return node
    return _canon_neg_value(node.args[1])
end

function _canon_neg_value(arg::Expr)
    if arg isa IntExpr
        return IntExpr(-arg.value)
    elseif arg isa NumExpr
        return NumExpr(-arg.value)
    elseif arg isa OpExpr && arg.op == "neg" && length(arg.args) == 1
        return arg.args[1]
    end
    return OpExpr("neg", Expr[arg])
end

function _flatten_same_op(args::Vector{Expr}, op::String)
    out = Expr[]
    for a in args
        if a isa OpExpr && a.op == op
            append!(out, a.args)
        else
            push!(out, a)
        end
    end
    return out
end

function _partition_identity(args::Vector{Expr}, identity::Int)
    others = Expr[]
    had_int = false
    had_float = false
    for a in args
        if a isa IntExpr && a.value == identity
            had_int = true
            continue
        end
        if a isa NumExpr && a.value == Float64(identity)
            had_float = true
            continue
        end
        push!(others, a)
    end
    return others, had_int, had_float
end

_all_float_literals(args::Vector{Expr}) = !isempty(args) && all(a -> a isa NumExpr, args)

_is_zero_any(e) = (e isa IntExpr && e.value == 0) || (e isa NumExpr && e.value == 0.0)
_is_one_any(e)  = (e isa IntExpr && e.value == 1) || (e isa NumExpr && e.value == 1.0)

function _arg_tier(e::Expr)
    if e isa IntExpr || e isa NumExpr
        return 0
    elseif e isa VarExpr
        return 1
    elseif e isa OpExpr
        return 2
    end
    return 3
end

_numeric_key(e::IntExpr) = Float64(e.value)
_numeric_key(e::NumExpr) = e.value
_numeric_key(e::Expr)    = 0.0

function _sort_args!(args::Vector{Expr})
    cache = Dict{Int,String}()
    function get_json(idx::Int, e::Expr)
        haskey(cache, idx) && return cache[idx]
        s = _emit_json(e)
        cache[idx] = s
        return s
    end
    indices = collect(1:length(args))
    sort!(indices; lt = (i, j) -> _compare(args[i], args[j], i, j, get_json))
    snap = [args[i] for i in indices]
    for i in 1:length(args)
        args[i] = snap[i]
    end
end

function _compare(a::Expr, b::Expr, ia::Int, ib::Int, get_json)
    ta, tb = _arg_tier(a), _arg_tier(b)
    if ta != tb
        return ta < tb
    end
    if ta == 0
        av, bv = _numeric_key(a), _numeric_key(b)
        if av != bv
            return av < bv
        end
        # int before float at equal magnitude.
        return (a isa IntExpr) && (b isa NumExpr)
    elseif ta == 1
        return (a::VarExpr).name < (b::VarExpr).name
    else
        return get_json(ia, a) < get_json(ib, b)
    end
end

"""
    canonical_json(expr::Expr) -> String

Emit the canonical on-wire JSON form of an expression per RFC §5.4.6: keys
sorted, no extraneous whitespace, shortest round-trip floats with trailing-`.0`
disambiguation for integer-valued magnitudes, exponent-form (`1e25`, `5e-324`)
outside `[1e-6, 1e21)`, signed zero preserved.
"""
function canonical_json(expr::Expr)::String
    return _emit_json(canonicalize(expr))
end

function _emit_json(e::Expr)::String
    if e isa IntExpr
        return string(e.value)
    elseif e isa NumExpr
        return format_canonical_float(e.value)
    elseif e isa VarExpr
        return _json_string(e.name)
    elseif e isa OpExpr
        return _emit_node_json(e)
    end
    error("cannot canonicalize value of type $(typeof(e))")
end

function _emit_node_json(n::OpExpr)::String
    entries = Tuple{String,String}[]
    push!(entries, ("op", _json_string(n.op)))
    args_str = "[" * join((_emit_json(a) for a in n.args), ",") * "]"
    push!(entries, ("args", args_str))
    if n.wrt !== nothing
        push!(entries, ("wrt", _json_string(n.wrt::String)))
    end
    if n.dim !== nothing
        push!(entries, ("dim", _json_string(n.dim::String)))
    end
    if n.name !== nothing
        push!(entries, ("name", _json_string(n.name::String)))
    end
    if n.value !== nothing
        push!(entries, ("value", _emit_canonical_value(n.value)))
    end
    sort!(entries, by = kv -> kv[1])
    body = join(("$(_json_string(k)):$v" for (k, v) in entries), ",")
    return "{" * body * "}"
end

function _json_string(s::AbstractString)::String
    # Use JSON3 to escape, but JSON3.write returns a string with surrounding
    # quotes. Strip and re-quote to avoid newlines/etc.
    return JSON3.write(String(s))
end

# Canonical JSON for `const`-op values. Values are JSON-typed (integer,
# Float64, AbstractString, or nested arrays thereof per esm-spec §4.2 / §9.2).
# Floats use the same canonical-float formatter as `NumExpr`; integers emit
# as bare digit-only tokens; arrays recurse element-wise.
function _emit_canonical_value(v)::String
    if v isa Bool
        return v ? "true" : "false"
    elseif v isa Integer
        return string(v)
    elseif v isa AbstractFloat
        return format_canonical_float(Float64(v))
    elseif v isa AbstractString
        return _json_string(v)
    elseif v isa AbstractArray
        return "[" * join((_emit_canonical_value(x) for x in v), ",") * "]"
    end
    throw(CanonicalizeError("E_CANONICAL_NONFINITE",
        "unsupported `const` value type: $(typeof(v))"))
end

"""
    format_canonical_float(f::Float64) -> String

Format a finite `Float64` per RFC §5.4.6.
"""
function format_canonical_float(f::Float64)::String
    isfinite(f) || throw(CanonicalizeError("E_CANONICAL_NONFINITE",
        "non-finite float in canonical form"))
    if f == 0.0
        return signbit(f) ? "-0.0" : "0.0"
    end
    a = abs(f)
    use_exp = a < 1e-6 || a >= 1e21
    if use_exp
        # Julia's default Base.show gives shortest round-trip ("Ryu"); for
        # scientific values it prints e.g. "1.0e25" or "5.0e-324". Need to
        # normalize: strip mantissa trailing zeros after '.', strip leading
        # exponent zeros, no leading + on exp.
        return _normalize_julia_exp(repr(f))
    end
    s = repr(f)
    if occursin('e', s) || occursin('E', s)
        # Within plain range but Julia chose exponent — convert to plain.
        s = _expand_to_plain(f)
    end
    if !occursin('.', s)
        s *= ".0"
    end
    return s
end

function _normalize_julia_exp(s::String)::String
    # Examples: "1.0e25" -> "1e25", "5.0e-324" -> "5e-324", "3.14e-10" -> "3.14e-10"
    s = lowercase(s)
    if !occursin('e', s)
        return s
    end
    parts = split(s, 'e'; limit=2)
    mant = parts[1]
    exp = parts[2]
    # Strip mantissa trailing zeros after '.'.
    if occursin('.', mant)
        mant = rstrip(mant, '0')
        mant = rstrip(mant, '.')
    end
    if isempty(mant)
        mant = "0"
    end
    # Normalize exponent: drop leading +, drop leading 0s preserving sign.
    if startswith(exp, "+")
        exp = exp[2:end]
    end
    sign = ""
    if startswith(exp, "-")
        sign = "-"
        exp = exp[2:end]
    end
    exp = lstrip(exp, '0')
    if isempty(exp)
        exp = "0"
    end
    return string(mant, "e", sign, exp)
end

function _expand_to_plain(f::Float64)::String
    # Julia's @sprintf with %.17g preserves precision; then trim trailing zeros.
    # Simpler: convert via string and special-case the rare path.
    s = repr(f)
    if !occursin('e', lowercase(s))
        return s
    end
    # Use BigFloat to get a clean expansion then trim.
    b = BigFloat(s)
    # Render without exponent — use printf format.
    out = string(b)
    # `string(BigFloat)` may still use scientific. Fallback: format manually.
    if occursin('e', lowercase(out))
        # Manual: split mantissa and exponent, shift decimal.
        parts = split(lowercase(s), 'e')
        mant = parts[1]
        exp = parse(Int, parts[2])
        out = _shift_decimal(mant, exp)
    end
    if occursin('.', out)
        out = rstrip(out, '0')
        out = rstrip(out, '.')
    end
    return out
end

function _shift_decimal(mant::AbstractString, exp::Int)::String
    sign = ""
    if startswith(mant, '-')
        sign = "-"
        mant = mant[2:end]
    end
    if !occursin('.', mant)
        mant = mant * "."
    end
    dot = findfirst('.', mant)
    digits = String(replace(String(mant), '.' => ""))
    pos = dot - 1 + exp  # new decimal position from start
    if pos <= 0
        digits = "0"^(1 - pos) * digits
        pos = 1
    elseif pos > length(digits)
        digits = digits * "0"^(pos - length(digits))
    end
    if pos == length(digits)
        return string(sign, digits)
    end
    return string(sign, digits[1:pos], ".", digits[pos+1:end])
end
